### DISCUSSION TITLE
[FYI] The latest development status

### DISCUSSION THREAD

**Discussion by eneim** [eneim/kohii#154] [General]
I would like to be more transparent about the latest status of this library, so it will be easier for anyone who is interested in this work to follow and consider contributing.

- Kohii was developed in nearly 3 years to reach 1.0 in March 2020, and it is about 3 hours every day. Since the beginning of 2020, I have become a full-time Android Engineer. As a result, time spent for Kohii is significantly reduced. Not only that I do not have a lot of time fixing bugs and implementing new features, but it is also harder to concentrate on the development overall. I'm not giving up, but it takes more time to deliver new code. I hope it would explain a lot about hanging issues and the slow release cycle.

- Moving forward, I want to address the following things:
  - Better state management. Currently, the Playable states are stored globally by the Master class. It causes the behavior where reopening Activity/Fragment doesn't reset the playback state as expected (e.g. https://github.com/eneim/kohii/issues/103). I'm investing in the solution of using ViewModel, which sounds to be the right direction. **Current status**: drafting at https://github.com/eneim/kohii/pull/151
  - Consider letting the Manager class construct the Playable instead of the Engine class (e.g. Kohii class). The reason is, a Playable instance, at last, should be managed by a Manager. An un-managed Playable is scheduled to be destroyed. In the event of recreation, it is complicated to judge that a Playable is being _abandoned_ by the Manager or not (because the Manager will be recreated). It would be easier to handle the configuration changes if we know who creates the Playable, and who is the last Manager to manage it. **Current status**: not started.
  - Supporting ExoPlayer 2.13.x. The library has evolved a lot since 2.11.x, and supporting it is a natural goal. It takes some time to consume the changes, to experiment with the new features, and to merge the benefit to this library, but I'm also working on that now. **Current status**: local experiments.
  - Supporting Jetpack Compose. As Jetpack Compose is becoming more stable, it would be nice if this library can play an important role in Video playback in the new area of Android development. I'm still at the beginning of learning Compose, so supporting it would be a big milestone for my studying as well. **Current status**: not started.
  - Less preset setup (or, in another word, the client needs to do more, and the library needs to do less). Kohii library comes with a quick configuration that helps you to start playing Video in just a few lines of code. It is nice from the client's perspective, but internally it needs to do a lot of things, and they are all so opinionated. I want to give more power in making decisions to the client of the library so that the library can evolve freely without the burden of keeping things simple. One of the motivations to do this comes from this article: https://jakewharton.com/integration-verbosity-and-good-layering/. **Current status**: not started.
 
- Further future of the library: I want to keep this library alive as long as possible. My challenge now is the time I have for it, and the experience I have to build a flexible codebase that does not only solve meaningful problems but is also easy to contribute. With more experience developing the Android library and relevant technologies, I hope to make it easier for contributors to join and support it.

**Comment by loicsiret** [comment]
If I can help, don't hesitate :)
